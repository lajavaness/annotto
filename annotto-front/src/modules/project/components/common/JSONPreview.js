import { isEmpty } from 'lodash'
import PropTypes from 'prop-types'
import React from 'react'

import { BOOLEAN, NUMBER, OBJECT, STRING } from 'shared/enums/rawTypes'

import * as Styled from 'modules/project/components/common/__styles__/JSONPreview.styles'

const JSONPreview = ({ json }) => {
	const highlightedJSON = (jsonObj) =>
		Object.keys(jsonObj).map((key) => {
			const value = jsonObj[key]
			let valueType = typeof value

			const isSimpleValue = [STRING, NUMBER, BOOLEAN].includes(valueType) || !value
			if (isSimpleValue && valueType === 'object') {
				valueType = 'null'
			}

			const resolveNotationStart = valueType === OBJECT ? (Array.isArray(value) ? '[' : '{') : ''
			const resolveNotationEnd = valueType === OBJECT ? (Array.isArray(value) ? ']' : '}') : ''

			return (
				<Styled.Container key={key}>
					<Styled.KeyLabel>{`${key}: ${resolveNotationStart}`}</Styled.KeyLabel>
					{!(valueType === OBJECT && isEmpty(value)) ? (
						isSimpleValue ? (
							<Styled.ValueLabel $valueType={valueType}>{JSON.stringify(value)}</Styled.ValueLabel>
						) : (
							<Styled.ValueContent>{highlightedJSON(value)}</Styled.ValueContent>
						)
					) : null}
					{`${resolveNotationEnd},`}
				</Styled.Container>
			)
		})

	return <Styled.Root>{highlightedJSON(json)}</Styled.Root>
}

export default JSONPreview

JSONPreview.propTypes = {
	/** The Json to display. */
	json: PropTypes.object,
}

JSONPreview.defaultProps = {
	json: {},
}
