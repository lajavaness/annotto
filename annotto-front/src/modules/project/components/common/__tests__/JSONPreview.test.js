import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'
import React from 'react'

import theme from '__theme__'

import JSONPreview from 'modules/project/components/common/JSONPreview'

const defaultProps = {}
const getInstance = (props = {}) => (
	<ThemeProvider theme={theme}>
		<JSONPreview {...defaultProps} {...props} />
	</ThemeProvider>
)

describe('JSONPreview', () => {
	it('matches snapshot', () => {
		const { asFragment } = render(getInstance())
		expect(asFragment()).toMatchSnapshot()
	})
})
