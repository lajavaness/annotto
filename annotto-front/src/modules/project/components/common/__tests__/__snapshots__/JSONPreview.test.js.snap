// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONPreview matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gsBqHK ixPyrw"
  />
</DocumentFragment>
`;
